below/ethtool/src/ethtool.rs

```diff
@@ -0,0 +1,234 @@
+    mask: u32,
+    data: usize,
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct GStrings {
+    pub cmd: u32,
+    pub string_set: u32,
+    pub len: u32,
+    pub data: [u8; MAX_GSTRINGS * ETH_GSTRING_LEN],
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct GStats {
+    pub cmd: u32,
+    pub len: u32,
+    pub data: [u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct IfReq {
+    if_name: [u8; IFNAME_MAX_SIZE],
+    if_data: usize,
+}
+
+fn _ioctl(fd: i32, if_name: &str, data: usize) -> Result<(), Errno> {
+    let mut ifname = [0u8; IFNAME_MAX_SIZE];
+    match ifname.get_mut(..if_name.len()) {
+        Some(slice) => slice.copy_from_slice(if_name.as_bytes()),
+        None => return Err(Errno::EINVAL),
+    };
+    let mut request = IfReq {
+        if_name: ifname,
+        if_data: data,
+    };
+
+    let exit_code = unsafe { ioctl(fd, nix::libc::SIOCETHTOOL, &mut request) };
+
+    if exit_code != 0 {
+        println!("code: {exit_code}, data: {:?}", request.if_data);
+        return Err(Errno::from_i32(exit_code));
+    }
+    Ok(())
+}
+
+/// Parses the byte array returned by ioctl for ETHTOOL_GSTRINGS command.
+/// In case of error during parsing any stat name,
+/// the function returns a `ParseError`.
+fn parse_names(
+    data: [u8; MAX_GSTRINGS * ETH_GSTRING_LEN],
+    length: usize,
+) -> Result<Vec<String>, Error> {
+    let names = data
+        .chunks(ETH_GSTRING_LEN)
+        .map(|chunk| {
+            // Find the position of the null terminator for specific stat name
+            let null_pos = chunk.iter().position(|b| *b == 0).unwrap_or(length);
+            // Convert the stat name to a string
+            str::from_utf8(&chunk[..null_pos])
+                .map(|s| s.to_string())
+                .map_err(|err| Error::ParseError(err.to_string()))
+        })
+        .collect::<Result<Vec<String>, Error>>()?;
+
+    Ok(names)
+}
+
+/// Parses the byte array returned by ioctl for ETHTOOL_GSTATS command.
+/// In case of error during parsing any feature,
+/// the function returns a `ParseError`.
+fn parse_values(
+    data: [u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+    length: usize,
+) -> Result<Vec<u64>, Error> {
+    let mut values = Vec::with_capacity(length);
+    let mut value_bytes = [0u8; 8];
+    for i in 0..length {
+        let offset = 8 * i;
+        match data.get(offset..offset + 8) {
+            Some(slice) => {
+                value_bytes.copy_from_slice(slice);
+                values.push(u64::from_ne_bytes(value_bytes));
+            }
+            None => {
+                return Err(Error::ParseError(format!(
+                    "parse value failed at offset={}",
+                    offset
+                )))
+            }
+        }
+    }
+
+    Ok(values)
+}
+
+pub trait EthtoolReadable {
+    fn new(if_name: &str) -> Result<Self, Error> where Self: Sized;
+    fn stats(&self) -> Result<Vec<(String, u64)>, Error>;
+}
+
+pub struct Ethtool {
+    sock_fd: i32,
```

Comment:
You should use https://doc.rust-lang.org/stable/std/os/fd/struct.OwnedFd.html so you don't need your own Drop impl